flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkBushyJoinReorderRule.java

```diff
@@ -0,0 +1,634 @@
+        // or no possible plan exists for next level.
+        while (foundPlans.size() < numJoinFactors) {
+            Map<Set<Integer>, JoinPlan> nextLevelJoinPlanMap =
+                    foundNextLevel(relBuilder, new ArrayList<>(foundPlans), multiJoin);
+            if (nextLevelJoinPlanMap.size() == 0) {
+                break;
+            }
+            foundPlans.add(nextLevelJoinPlanMap);
+        }
+
+        return foundPlans;
+    }
+
+    private static boolean outerJoinConditionExists(LoptMultiJoin multiJoin) {
+        for (int i = 0; i < multiJoin.getNumJoinFactors(); i++) {
+            if (multiJoin.getOuterJoinCond(i) != null
+                    && RelOptUtil.conjunctions(multiJoin.getOuterJoinCond(i)).size() != 0) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static JoinPlan getBestPlan(Map<Set<Integer>, JoinPlan> levelPlan) {
+        JoinPlan bestPlan = null;
+        for (Map.Entry<Set<Integer>, JoinPlan> entry : levelPlan.entrySet()) {
+            if (bestPlan == null || entry.getValue().betterThan(bestPlan)) {
+                bestPlan = entry.getValue();
+            }
+        }
+
+        return bestPlan;
+    }
+
+    private static JoinPlan addToTopForOuterJoin(
+            JoinPlan bestPlan, LoptMultiJoin multiJoin, RelBuilder relBuilder) {
+        List<Integer> remainIndexes = new ArrayList<>();
+        for (int i = 0; i < multiJoin.getNumJoinFactors(); i++) {
+            if (!bestPlan.factorIds.contains(i)) {
+                remainIndexes.add(i);
+            }
+        }
+
+        RelNode leftNode = bestPlan.relNode;
+        LinkedHashSet<Integer> set = new LinkedHashSet<>(bestPlan.factorIds);
+        for (int index : remainIndexes) {
+            RelNode rightNode = multiJoin.getJoinFactor(index);
+
+            // Make new join condition and get new join type.
+            Optional<Tuple2<Set<RexCall>, JoinRelType>> joinConditions =
+                    getConditionsAndJoinType(
+                            bestPlan.factorIds, Collections.singleton(index), multiJoin, true);
+
+            if (!joinConditions.isPresent()) {
+                continue;
+            } else {
+                // Is left/right outer join, but we all given left join type.
+                Set<RexCall> conditions = joinConditions.get().f0;
+                List<RexNode> rexCalls = new ArrayList<>(conditions);
+                Set<RexCall> newCondition =
+                        convertToNewCondition(
+                                new ArrayList<>(set),
+                                Collections.singletonList(index),
+                                rexCalls,
+                                multiJoin);
+
+                leftNode =
+                        relBuilder
+                                .push(leftNode)
+                                .push(rightNode)
+                                .join(JoinRelType.LEFT, newCondition)
```

Comment:
> Why is it always `JoinRelType.LEFT` here ?

As a simple case for example. `((A RJ B) IJ C)`, the left side  join condition of `(A RJ B) IJ C` comes from `B`.  This case can be converted to `MultiJoin` whose join type in `MultiJoin` is `[Right, Inner, Inner]`. So, In our bushyJoinReoderRule, we will first reorder `B` and `C` to form `B IJ C`. Then, we will add `A` to the top, and it join type need to be converted to left join. So, for each left/right join, if it can be added into a `MultiJoin`, it will be converted to left join.

Reply:
> Why is it always `JoinRelType.LEFT` here ?

The join type of left join : `((A LJ B) IJ C)` in `MultiJoin` is `[Inner, Left, Inner]`.